### ISSUE TITLE
Should there be a warning if limit is reached?

### ISSUE THREAD

**Issue by joethorley** [poissonconsulting/fwatlasbc#20]

**Comment by sebdalgarno** [comment]
an alternative would be to make the default limit a very large number

**Comment by joethorley** [comment]
Is infinity a possibility?

**Comment by sebdalgarno** [comment]
perhaps....is that better than .max_integer?

**Comment by joethorley** [comment]
The intent is clearer

**Comment by sebdalgarno** [comment]
changing default to NULL (assuming that would return all rows regardless of size) had some unexpected behaviour... most tests passed but a couple failed (i.e. returned objects had differing number of rows than when limit was 1000 even though both cases were well below that, e.g. 30 vs 89). I don't fully understand what limit = NULL does in this case so don't think that is a good solution. 

I've also realised that the API doesn't accept limit values >= 100,000. The error returned from the API is: "Invalid value for parameter limit: 1e+05"

More confusing, when I try any limit > 10,000 but <100,000 no error is returned but the returned object has 10,000 rows...which seems wrong unless the fwa_stream_networks_sp layer has exactly 10,000 rows. So it seems that the API is imposing a limit of 10,000 regardless of the value provided. 

@smnorris do you have any thoughts on this?

**Comment by smnorris** [comment]
Yes, the hard limit per request is 10k, you have to make another request to get more records:
https://access.crunchydata.com/documentation/pg_featureserv/latest/installation/configuration/

The default was set at crazy low 20, I bumped it up to be equivalent to the max. And the province sets their WFS limit at 10k as well so I stuck with that.
```
[Paging]
# The default number of features in a response
LimitDefault = 10000
# Maximum number of features in a response
LimitMax = 10000
```

We could experiment with higher maximums, I'm not sure what factors come into play.
But it does make sense to me that the user should break really big requests into smaller pieces. 

There are probably better ways to deal with paging, but what I do with `bcdata` is find out how many features I want, prep all the requests with the various limits and then make the requests in parallel. https://github.com/smnorris/bcdata/blob/master/bcdata/wfs.py#L253

**Comment by smnorris** [comment]
It looks like `limit` and `offset` are what you want:
https://access.crunchydata.com/documentation/pg_featureserv/1.1.1/usage/query_data/

The OGC API spec http://docs.opengeospatial.org/is/17-069r3/17-069r3.html does mention paging links in responses, and so does the WFS spec.
I'm not sure if pg_featureserv provides links like that - and I was never able to find paging links in the BC WFS either.

**Comment by sebdalgarno** [comment]
Great @smnorris thanks that clears it up. @joethorley I see two options going forward:
1. make default limit 10,000 and expose the offset argument from fwapgr to allow the user to do paging themselves
2. make default limit some very large number (or infinity or NULL for all of them) and do the paging ourselves within the functions. This would mean the offset argument remains hidden to the user but we could notify them if paging occurs. 

One question is how often will a user require >10,000 records and would they be better off using fwapgr? Given our watershed-based workflow is there a scenario where there are >10,000 streams in a watershed? Seems to me the likely scenario for that many features is to get an entire collection (i.e. the whole fwa_stream_networks_sp dataset).

**Comment by joethorley** [comment]
Yes I think 1 is the best option. We may end up developing a higher level package that does the paging.

**Comment by sebdalgarno** [comment]
note for myself - there is unexpected behaviour with offset argument
```
x <- fwa_get_stream_network(wshed, limit = 10000, offset = 0)
```
returns an object with 89 rows
```
x <- fwa_get_stream_network(wshed, limit = 10000, offset = 50)
```
returns an object with 88 rows although I expected 39 rows

need to look into whether the same behaviour occurs when using the API directly

**Comment by sebdalgarno** [comment]
re: provding message to user when limit is reached:

from Joe:
I’m thinking its probably worth it because if it just trims off the end of some data folks may never notice. There could be a quiet or silent argument with a global option for folks to switch on for the package but it should be off by default ie quiet = getOption(fwa.quiet, FALSE).
Note It seems to me that quiet to suppress messages and silent to suppress messages and warnings would be a reasonable convention?

**Comment by sebdalgarno** [comment]
this should be a warning
